app/code/Magento/Reports/Test/Unit/Observer/EventSaverTest.php

```diff
@@ -0,0 +1,156 @@
+
+    /**
+     * @var EventSaver
+     */
+    private $eventSaver;
+
+    /**
+     * @inheritDoc
+     */
+    public function setUp()
+    {
+        $this->customerSessionMock = $this->createMock(Session::class);
+        $this->customerVisitorMock = $this->createMock(Visitor::class);
+
+        $this->storeMock = $this->createMock(StoreInterface::class);
+        $this->storeManagerMock = $this->createMock(StoreManagerInterface::class);
+        $this->storeManagerMock->expects($this->once())
+            ->method('getStore')
+            ->willReturn($this->storeMock);
+        $this->storeMock->expects($this->once())
+            ->method('getId')
+            ->willReturn(1);
+
+        $this->eventMock = $this->createMock(Event::class);
+        $this->eventFactoryMock = $this->createMock(EventFactory::class);
+        $this->eventFactoryMock->expects($this->once())
+            ->method('create')
+            ->willReturn($this->eventMock);
+        $this->eventMock->expects($this->once())
+            ->method('setData')
+            ->willReturnSelf();
+        $this->eventMock->expects($this->once())
+            ->method('save')
+            ->willReturnSelf();
+
+        $objectManagerHelper = new ObjectManager($this);
+        $this->eventSaver = $objectManagerHelper->getObject(
+            EventSaver::class,
+            [
+                '_storeManager' => $this->storeManagerMock,
+                '_eventFactory' => $this->eventFactoryMock,
+                '_customerSession' => $this->customerSessionMock,
+                '_customerVisitor' => $this->customerVisitorMock
+            ]
+        );
+    }
+
+    /**
+     * Test save with subject ID provided
+     */
+    public function testSaveWithSubject()
+    {
+        $subjectId = 5;
+        $this->customerSessionMock->expects($this->never())
+            ->method('isLoggedIn');
+        $this->customerSessionMock->expects($this->never())
+            ->method('getCustomerId');
+        $this->customerVisitorMock->expects($this->never())
+            ->method('getId');
+        $this->eventSaver->save(1, 1, $subjectId);
+    }
+
+    /**
+     * Test save with no subject ID provided
+     *
+     * @param bool $isLoggedIn
+     * @dataProvider dataProviderTestSaveWithoutSubject
+     */
+    public function testSaveWithoutSubject(bool $isLoggedIn)
+    {
+        if ($isLoggedIn) {
+            $this->customerSessionMock->expects($this->once())
+                ->method('isLoggedIn')
+                ->willReturn(true);
+            $this->customerSessionMock->expects($this->once())
+                ->method('getCustomerId')
+                ->willReturn(1);
+        } else {
+            $this->customerSessionMock->expects($this->once())
+                ->method('isLoggedIn')
+                ->willReturn(false);
+            $this->customerVisitorMock->expects($this->once())
+                ->method('getId')
+                ->willReturn(2);
+        }
```

Comment:
Split that into 2 separate tests:
* `testSaveWithoutSubjectWhenLoggedIn`
* `testSaveWithoutSubjectForLoggedIn`

Reply:
thanks for the comment, done.